tests/pos/i2944.scala

```diff
@@ -0,0 +1,10 @@
+trait Map2[K] {
+  def get(k: K): K = k
+  def foo: K = {
+    this match {
+      case that: Map2[b] => that.get(3.asInstanceOf[b])
+      //case that: Map2[c] => that.get(4.asInstanceOf[K])
```

Comment:
This one still crashes while trying to type `that.get(3.asInstanceOf[K].asInstanceOf[b @ b])`.